File: LeetCode/2610-closest-prime-numbers-in-range/closest-prime-numbers-in-range.go
func closestPrimes(left int, right int) []int {
    const N = 1e6 + 1
    isPrime := make([]bool, N)
    primes := []int{}

    for i := range isPrime{
        isPrime[i] = true
    }

    isPrime[0], isPrime[1] = false, false

    for i := 2; i*i <= int(N); i++ {
		if isPrime[i] {
			for j := i * i; j < int(N); j += i {
				isPrime[j] = false
			}
		}
	}

    for i := 2; i < int(N); i++ {
		if isPrime[i] {
			primes = append(primes, i)
		}
	}

    low := sort.Search(len(primes), func(i int) bool {return primes[i] >= left})
    high := sort.Search(len(primes), func(i int) bool {return primes[i] > right}) - 1

    if len(primes) < 2{
        return []int{-1, -1}
    }

    num1, num2, diff := -1, -1, math.MaxInt32
    for i := low + 1; i <= high; i++{
        if primes[i]-primes[i-1] < diff {
			diff = primes[i] - primes[i-1]
			num1 = primes[i-1]
			num2 = primes[i]
		}
    }

    return []int{num1, num2}
}

File: LeetCode/2610-closest-prime-numbers-in-range/closest-prime-numbers-in-range.cpp
class Solution {
public:
    vector<int> closestPrimes(int left, int right) {
        vector<int> isPrime(1e6+1, 1);
        vector<int> primes;

        isPrime[0] = isPrime[1] = 0;
        for(int i = 2; i * i <= 1e6; i++){
            if(isPrime[i]){
                for(int j = i * i; j <= 1e6; j += i){
                    isPrime[j] = false;
                }
            }
        }

        for(int i = 2; i <= 1e6; i++){
            if(isPrime[i])
                primes.push_back(i);
        }

        int low = lower_bound(primes.begin(), primes.end(), left) - primes.begin();
        int high = upper_bound(primes.begin(), primes.end(), right) - primes.begin() - 1;

        if(primes.size() < 2)
            return {-1, -1};

        int num1 = -1, num2 = -1;
        int diff = INT_MAX;
        for(int i = low + 1; i <= high; i++){
            if(primes[i] - primes[i-1] < diff){
                diff = primes[i] - primes[i-1];
                num1 = primes[i-1];
                num2 = primes[i];
            }
        }

        return {num1, num2};
    }
};